common/djangoapps/third_party_auth/pipeline.py

```diff
@@ -470,63 +445,56 @@ def ensure_user_information(strategy, auth_entry, user=None, *args, **kwargs):
             return dispatch_to_register()
-        elif user_inactive:
-            # If the user has a linked account, but has not yet activated
-            # we should send them to the login page. The login page
-            # will tell them that they need to activate their account.
-            return dispatch_to_login()
-
-
-def _create_redirect_url(url, strategy):
-    """ Given a URL and a Strategy, construct the appropriate redirect URL.
-
-    Construct a redirect URL and append the URL parameters that should be preserved.
-
-    Args:
-        url (string): The base URL to use for the redirect.
-        strategy (Strategy): Used to determine which URL parameters to append to the redirect.
-
-    Returns:
-        A string representation of the URL, with parameters, for redirect.
-    """
-    url_params = {}
-    enroll_course_id = strategy.session_get(AUTH_ENROLL_COURSE_ID_KEY)
-    if enroll_course_id:
-        url_params['course_id'] = enroll_course_id
-        url_params['enrollment_action'] = 'enroll'
-    email_opt_in = strategy.session_get(AUTH_EMAIL_OPT_IN_KEY)
-    if email_opt_in:
-        url_params[AUTH_EMAIL_OPT_IN_KEY] = email_opt_in
-    if url_params:
-        return u'{url}?{params}'.format(
-            url=url,
-            params=urllib.urlencode(url_params)
-        )
-    else:
-        return url
+        elif auth_entry == AUTH_ENTRY_ACCOUNT_SETTINGS:
+            raise AuthEntryError(backend, 'auth_entry is wrong. Settings requires a user.')
+        else:
+            raise AuthEntryError(backend, 'auth_entry invalid')
+
+    if not user.is_active:
+        # The user account has not been verified yet.
+        if allow_inactive_user:
+            # This parameter is used by the auth_exchange app, which always allows users to
+            # login, whether or not their account is validated.
+            pass
+        # IF the user has just registered a new account as part of this pipeline, that is fine
+        # and we allow the login to continue this once, because if we pause again to force the
+        # user to activate their account via email, the pipeline may get lost (e.g. email takes
+        # too long to arrive, user opens the activation email on a different device, etc.).
+        # This is consistent with first party auth and ensures that the pipeline completes
+        # fully, which is critical.
+        # But if this is an existing account, we refuse to allow them to login again until they
+        # check their email and activate the account.
+        elif social is not None:
+            # This third party account is already linked to a user account. That means that the
+            # user's account existed before this pipeline originally began (since the creation
+            # of the 'social' link entry occurs in one of the following pipeline steps).
+            # Reject this login attempt and tell the user to validate their account first.
+
+            # TODO: resend validation email
```

Comment:
The existing SSO implementation resends this activation email. I'd consider excluding that from this work a regression.